### PR TITLE
Local exceptions (without the optimization)

### PR DESCRIPTION
**Pull request by alainfrisch**
This is a stripped down version of #260.  Here I only propose to add local exception definitions, with no specific compilation (which can come later or not).

Local exceptions should be encouraged, since they document a more limited scope, which helps reasoning about the code and avoid subtle bug when higher-order function use a generic exception (such as Exit) for their internal control flow but can easily get confused if they call unknown function which could also raise the same exception.

Moreover, now that it is no longer allowed to define multiple exceptions with the same name in a given structure, it becomes even more important not to pollute the global scope.

There is of course a work-around for encoding local exceptions with local modules, but it's syntactically heavier (two names need to be invented) and less efficient (unless more optimization are added).  People would more be likely to use global exceptions if we don't provide something lighter than this encoding.

See https://blogs.janestreet.com/a-universal-type/ for some fun use of (previously encoded) local exceptions that would benefit from this feature:

``` ocaml
module type Univ = sig 
  type t 
  val embed: unit -> ('a -> t) * (t -> 'a option) 
end 

module U : Univ = struct
  type t = exn

  let embed (type s) () =
    let exception E of s in
   (fun x -> E x), (function E x -> Some x | _ -> None)
end
```

Note: the grammar has been massaged a little bit to avoid a conflict with `exception` patterns:

``` ocaml
let exception Exit = ... in ...
let exception Exit in ...
```

As a result, exception patterns are now disallowed in let bindings. They were rejected by the type-checker anyway.

### CODE REVIEW

File: parsing/parser.mly [hunk 1]
```diff
@@ -1513,7 +1522,7 @@ let_binding_body:
   | val_ident COLON TYPE lident_list DOT core_type EQUAL seq_expr
       { let exp, poly = wrap_type_annotation $4 $6 $8 in
         (ghpat(Ppat_constraint(mkpatvar $1 1, poly)), exp) }
-  | pattern EQUAL seq_expr
+  | pattern_no_exn EQUAL seq_expr
```
Comment (damiendoligez) [inline parsing/parser.mly]: Looks like you've mixed up this patch with an unrelated change?

**Comment by damiendoligez** [comment]
FTR, I'm strongly in favour of the language change, but I don't feel qualified to review the implementation in detail.

**Comment by alainfrisch** [comment]
> I'm strongly in favour of the language change

Including for 4.03?

**Comment by damiendoligez** [comment]
> Including for 4.03?

We don't have a good consensus yet. What are the drawbacks of delaying it to 4.04?

**Comment by alainfrisch** [comment]
> What are the drawbacks of delaying it to 4.04?

Just extra work to maintain the list of active open PR and go back to it later.  I don't think it would be technically difficult to maintain the branch synchronized with trunk.

**Comment by yakobowski** [comment]
> What are the drawbacks of delaying it to 4.04?

People using old versions of OCaml will get even less chance to use this new syntax before dying of old age? :sunglasses:

**Comment by alainfrisch** [comment]
Rebased to trunk.

From the comments above, @damiendoligez is strongly in favor, and I suspect @yakobowski as well.  Is anyone opposed to merging to trunk?

**Comment by damiendoligez** [comment]
IIRC, the only opposition was from @lefessan, everyone else liked it.

Can you find someone to review the code?

**Comment by alainfrisch** [comment]
@garrigue Do you think you could have a look at the type-checking part of this PR?  (This is less than 10 lines, the rest is syntactic boilerplate or moving code around.)

**Comment by garrigue** [comment]
This looks fine to me. But some tests would help confidence.

**Comment by alainfrisch** [comment]
Now complete with a non-regression test, an entry in Changes and in the manual.
